TeamCode/src/main/java/com/acmerobotics/relicrecovery/path/LinearPath.java

```diff
@@ -0,0 +1,216 @@
+ * Created by Ryan on 4/5/2017.
+ */
+public class LinearPath {
+
+    public static final double EPSILON = 0.0001;
+
+    public static class DistanceReport {
+        public Segment segment;
+        public Vector2d queryPoint, pathPoint;
+        public double distance;
+    }
+
+    public static class Segment {
+        private Vector2d start, end, seg;
+
+        public Segment(Vector2d start, Vector2d end) {
+            this.start = start;
+            this.end = end;
+            this.seg = this.start.negated().add(this.end);
+        }
+
+        public Vector2d start() {
+            return start;
+        }
+
+        public Vector2d end() {
+            return end;
+        }
+
+        public Vector2d seg() {
+            return seg;
+        }
+
+        public double length() {
+            return this.seg.norm();
+        }
+
+        /** @param t range [0, 1] inclusive */
+        public Vector2d getPoint(double t) {
+            return this.seg.multiplied(t).add(this.start);
+        }
+
+        public Vector2d getBoundedPoint(double t) {
+            if (t <= 0) {
+                return this.start.copy();
+            } else if (t >= 1) {
+                return this.end.copy();
+            } else {
+                return getPoint(t);
+            }
+        }
+
+        /** @return [0, 1] position on curve; NaN if not on curve */
+        public double getPosition(Vector2d point) {
+            Vector2d adj = this.start.negated().add(point);
+            double tX = adj.x() / seg.x();
+            double tY = adj.y() / seg.y();
+            if (Math.abs(seg.x()) < EPSILON) {
+                return tY;
+            } else if (Math.abs(seg.y()) < EPSILON) {
+                return tX;
+            } else if (Math.abs(tX - tY) < EPSILON) {
+                return (tX + tY) / 2.0;
+            } else {
+                return Double.NaN;
+            }
+        }
+
+        public boolean contains(Vector2d point) {
+            double pos = getPosition(point);
+            return !Double.isNaN(pos) && pos >= 0 && pos <= 1;
+        }
+
+        public double getClosestPositionOnPath(Vector2d point) {
+            double a = seg.dot(start.negated().add(point));
+            return a / seg.dot(seg);
+        }
+
+        public static double getDistance(Vector2d a, Vector2d b) {
+            return a.negated().add(b).norm();
+        }
+
+        public boolean equals(Segment s) {
+            return s.start.equals(start) && s.end.equals(end);
+        }
+
+        @Override
+        public String toString() {
+            return String.format(Locale.ENGLISH, "<%f, %f> to <%f, %f>", start.x(), start.y(), end.x(), end.y());
+        }
+
+        public String getEquation() {
+            return String.format(Locale.ENGLISH, "<%f + %ft, %f + %ft>", start.x(), seg.x(), start.y(), seg.y());
+        }
+    }
+
+    private List<Segment> segments;
+    private double length;
+
+    public LinearPath(List<Vector2d> waypoints) {
```

Comment:
I kept `Vector2d` as the basis for this class. This is similar to what 254 uses (they use the functionally equivalent `Translation2d`). Nevertheless, it might be good to change this to `Pose2d`.

Reply:
If we want to be able to combine transations and rotations that would be important.

Reply:
Sounds good. I'll modify things to use `Pose2d`s